src/Type/Php/ArrayFilterFunctionReturnTypeExtension.php

```diff
@@ -71,69 +81,62 @@ public function getTypeFromFunctionCall(FunctionReflection $functionReflection,
+				if ($mode === self::USE_ITEM) {
+					$keyVar = null;
+					$itemVar = $callbackArg->params[0]->var;
+				} elseif ($mode === self::USE_KEY) {
+					$keyVar = $callbackArg->params[0]->var;
+					$itemVar = null;
+				} elseif ($mode === self::USE_BOTH) {
+					$keyVar = $callbackArg->params[1]->var ?? null;
+					$itemVar = $callbackArg->params[0]->var;
 				}
-			} elseif ($callbackArg instanceof ArrowFunction && count($callbackArg->params) > 0) {
-				return $this->filterByTruthyValue($scope, null, $arrayArgType, $callbackArg->params[0]->var, $callbackArg->expr);
-			} elseif ($callbackArg instanceof String_) {
-				$funcName = self::createFunctionName($callbackArg->value);
-				if ($funcName === null) {
-					return new ErrorType();
-				}
-
-				$keyVar = new Variable('key');
-				$expr = new FuncCall($funcName, [new Arg($keyVar)]);
-				return $this->filterByTruthyValue($scope, null, $arrayArgType, $keyVar, $expr);
+				return $this->filterByTruthyValue($scope, $itemVar, $arrayArgType, $keyVar, $statement->expr);
 			}
-		}
-
-		if ($flagArg instanceof ConstFetch && $flagArg->name->getParts()[0] === 'ARRAY_FILTER_USE_BOTH') {
-			if ($callbackArg instanceof Closure && count($callbackArg->stmts) === 1 && count($callbackArg->params) > 0) {
-				$statement = $callbackArg->stmts[0];
-				if ($statement instanceof Return_ && $statement->expr !== null) {
-					return $this->filterByTruthyValue($scope, $callbackArg->params[0]->var, $arrayArgType, $callbackArg->params[1]->var ?? null, $statement->expr);
-				}
-			} elseif ($callbackArg instanceof ArrowFunction && count($callbackArg->params) > 0) {
-				return $this->filterByTruthyValue($scope, $callbackArg->params[0]->var, $arrayArgType, $callbackArg->params[1]->var ?? null, $callbackArg->expr);
-			} elseif ($callbackArg instanceof String_) {
-				$funcName = self::createFunctionName($callbackArg->value);
+		} elseif ($callbackArg instanceof ArrowFunction && count($callbackArg->params) > 0) {
+			if ($mode === self::USE_ITEM) {
+				$keyVar = null;
+				$itemVar = $callbackArg->params[0]->var;
+			} elseif ($mode === self::USE_KEY) {
+				$keyVar = $callbackArg->params[0]->var;
+				$itemVar = null;
+			} elseif ($mode === self::USE_BOTH) {
+				$keyVar = $callbackArg->params[1]->var ?? null;
+				$itemVar = $callbackArg->params[0]->var;
+			}
+			return $this->filterByTruthyValue($scope, $itemVar, $arrayArgType, $keyVar, $callbackArg->expr);
+		} elseif (
+			($callbackArg instanceof FuncCall || $callbackArg instanceof MethodCall || $callbackArg instanceof StaticCall)
+			&& $callbackArg->isFirstClassCallable()
+		) {
+			[$args, $itemVar, $keyVar] = $this->createDummyArgs($mode);
+			$expr = clone $callbackArg;
+			$expr->args = $args;
+			return $this->filterByTruthyValue($scope, $itemVar, $arrayArgType, $keyVar, $expr);
+		} else {
+			$constantStrings = $scope->getType($callbackArg)->getConstantStrings();
+			if (count($constantStrings) === 1) {
```

Comment:
Can't you filter by all constant strings (and union the results?)

Testcase:

$fun = 'is_even';
if (rand(0,1)) $fun = 'is_int';

array_filter($array, $fun)

Reply:
Done.